### PR TITLE
[MERGE in 5.x] fix(compiler): check if entry components are reachable from their module

### PR DESCRIPTION
**Pull request by vicb**
**DO NOT MERGE IN THE 4.X BRANCH**

fixes #17466

### CODE REVIEW

File: packages/compiler/src/aot/compiler.ts [hunk 1]
```diff
@@ -440,6 +440,35 @@ function _analyzeNgModules(
...
+          mod: cmpModule.type.reference,
+        });
+      }
+    });
+
+    if (errors.length) {
+      const module = ngModuleMeta.type.reference;
+      const messages = errors.map(
+          error =>
+              `The module ${module.name} (${module.filePath}) uses the entry component ${error.cmp.name} from the module ${error.mod.name}.\n` +
```
Comment (chuckjaz) [inline packages/compiler/src/aot/compiler.ts]: Consider 

${module.filePath}: The component ${error.cmp.name}, declared in ${error.mod.name}, is declared as an entry component in module ${module.name) which does not import ${error.mod.name}. Consider adding ${error.mod.name} to the imports of ${module.name}. Note that ${error.cmp.name} might be referenced by a route which will automatically declare it as an entry component.

**Review by chuckjaz — APPROVED** [review]

**Comment by vicb** [comment]
This PR could actually lead to false positive: consider the use case described in the Router docs: https://angular.io/tutorial/toh-pt5#refactor-routes-to-a-routing-module

`DashboardComponent`, `HeroDetailComponent`, and `HeroesComponent` would be added as Entry component into `AppRoutingModule`. 

Because they are not declared in this module, the changes in this PR would result in an error being generated.

However bootstraping the `AppComponent` from the `AppModule` would work fine.

**Comment by vicb** [comment]
green presubmit in tap/160169163

**Comment by tbosch** [comment]
Maybe another way of solving this is during runtime:
- store the module class in every generated component factory.
- when the def of an `NgModuleFactory` is created (only 1x per module class), loop over the component factories that are stored in the `ComponentFactoryResolver` definition and check if the corresponding module class is part of the providers of the module.

As we are indexing providers via a Map, this should just be a linear walk of the component factories, so should be very cheap performance wise.

**Pull request closed without merging**